launcher/logs/LogParser.cpp

```diff
@@ -0,0 +1,366 @@
+                        // no op
+                    }
+                }
+                if (m_parser.hasError()) {
+                    break;
+                }
+            }
+
+            isCompleteLog4j = depth == 0;
+        }
+    }
+
+    if (isCompleteLog4j) {
+        return parseLog4J();
+    } else {
+        if (isPotentialLog4JStart(m_buffer)) {
+            m_partialData = QString(m_buffer);
+            return LogParser::Partial{ QString(m_buffer) };
+        }
+
+        int start = 0;
+        auto bufView = QStringView(m_buffer);
+        while (start < bufView.length()) {
+            if (qsizetype pos = bufView.right(bufView.length() - start).indexOf('<'); pos != -1) {
+                auto slicestart = start + pos;
+                auto slice = bufView.right(bufView.length() - slicestart);
+                if (isPotentialLog4JStart(slice)) {
+                    if (slicestart > 0) {
+                        auto text = m_buffer.left(slicestart);
+                        m_buffer = m_buffer.right(m_buffer.length() - slicestart);
+                        if (!text.trimmed().isEmpty()) {
+                            return LogParser::PlainText{ text };
+                        }
+                    }
+                    m_partialData = QString(m_buffer);
+                    return LogParser::Partial{ QString(m_buffer) };
+                }
+                start = slicestart + 1;
+            } else {
+                break;
+            }
+        }
+
+        // no log4j found, all plain text
+        auto text = QString(m_buffer);
+        m_buffer.clear();
+        if (text.trimmed().isEmpty()) {
+            return {};
+        } else {
+            return LogParser::PlainText{ text };
+        }
+    }
+}
+
+QList<LogParser::ParsedItem> LogParser::parseAvailable()
+{
+    QList<LogParser::ParsedItem> items;
+    bool doNext = true;
+    while (doNext) {
+        auto item_ = parseNext();
+        if (m_error.has_value()) {
+            return {};
+        }
+        if (item_.has_value()) {
+            auto item = item_.value();
+            if (std::holds_alternative<LogParser::Partial>(item)) {
+                break;
+            } else {
+                items.push_back(item);
+            }
+        } else {
+            doNext = false;
+        }
+    }
+    return items;
+}
+
+std::optional<LogParser::ParsedItem> LogParser::parseLog4J()
+{
+    m_parser.clear();
+    m_parser.setNamespaceProcessing(false);
+    m_parser.addData(m_buffer);
+
+    m_parser.readNextStartElement();
+    if (m_parser.qualifiedName() == "log4j:Event"_L1) {
+        auto entry_ = parseAttributes();
+        if (!entry_.has_value()) {
+            setError();
+            return {};
+        }
+        auto entry = entry_.value();
+
+        bool foundMessage = false;
+        int depth = 1;
+
+        while (!m_parser.atEnd()) {
+            auto tok = m_parser.readNext();
+            switch (tok) {
```

Comment:
I would have used lambdas to take a few of these switch cases more so because you have multiple switches inside each other